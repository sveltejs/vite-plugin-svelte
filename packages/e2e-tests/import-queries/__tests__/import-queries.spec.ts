import { browserLogs, fetchFromPage, getText, isBuild, testDir } from '~utils';
import { createServer, ViteDevServer } from 'vite';

function normalizeSnapshot(result: string) {
	// during dev, the import is rewritten but can vary on the v= hash. replace with stable short import
	return result
		.replace(/\.js\?v=[0-9a-f]{8}/g, '.js?v=XXX') // vite import analysis import rewrite version query
		.replace(/generated by Svelte v\d\.\d+\.\d+(?:-[a-z]+\.\d+)?/g, 'generated by Svelte vXXX') // compiler version comment
		.replace(/"total": *\d+\.\d+/g, '"total":0.123456789'); // svelte compile stats
}

describe('raw', () => {
	test('does not have failed requests', async () => {
		browserLogs.forEach((msg) => {
			expect(msg).not.toMatch('404');
		});
	});

	test('Dummy.svelte', async () => {
		expect(await getText('#component button')).toBe('dummy clicks: 0');
	});

	test('Dummy.svelte?raw', async () => {
		const result = await getText('#raw');
		expect(result).toMatchSnapshot();
	});

	test('Dummy.svelte?raw&svelte&type=preprocessed', async () => {
		const result = await getText('#preprocessed');
		expect(result).toMatchSnapshot();
	});

	test(`Dummy.svelte?raw&svelte&type=script`, async () => {
		const result = await getText('#script');
		expect(normalizeSnapshot(result)).toMatchSnapshot();
	});

	test('Dummy.svelte?raw&svelte&type=script&compilerOptions={"customElement":true}', async () => {
		const result = await getText('#wcScript');
		expect(normalizeSnapshot(result)).toMatchSnapshot();
	});

	test('Dummy.svelte?raw&svelte&type=style', async () => {
		const result = await getText('#style');
		expect(result).toMatchSnapshot();
	});

	test('Dummy.svelte?raw&svelte&type=all&sourcemap', async () => {
		const result = await getText('#all');
		expect(normalizeSnapshot(result)).toMatchSnapshot();
	});

	describe.runIf(!isBuild)('mixed exports', () => {
		test('Dummy.svelte?raw&svelte&type=preprocessed', async () => {
			const module = await fetchFromPage('src/Dummy.svelte?raw&svelte&type=preprocessed').then(
				(res) => res.text()
			);
			expect(normalizeSnapshot(module)).toMatchSnapshot();
		});
		test('Dummy.svelte?raw&svelte&type=style', async () => {
			const module = await fetchFromPage('src/Dummy.svelte?raw&svelte&type=style').then((res) =>
				res.text()
			);
			expect(normalizeSnapshot(module)).toMatchSnapshot();
		});
		test('Dummy.svelte?raw&svelte&type=script', async () => {
			const module = await fetchFromPage('src/Dummy.svelte?raw&svelte&type=script').then((res) =>
				res.text()
			);
			expect(normalizeSnapshot(module)).toMatchSnapshot();
		});
		test('Dummy.svelte?raw&svelte&type=all', async () => {
			const module = await fetchFromPage('src/Dummy.svelte?raw&svelte&type=all').then((res) =>
				res.text()
			);
			expect(normalizeSnapshot(module)).toMatchSnapshot();
		});
	});
});

// vitest prints a warning about obsolete snapshots during build tests, ignore it, they are used in dev tests.
// always regenerate snapshots with `pnpm test:serve import-queries -u` and check the diffs if they are correct
describe.runIf(isBuild)('snapshots not obsolete warning', async () => {
	afterAll(() => {
		console.log(
			'Ignore the obsolete snapshot warnings for ssrLoadModule snapshots from vitest during test:build, they are used in test:serve'
		);
	});
	test('suite not empty', () => {
		expect(true).toBe(true);
	});
});

describe.runIf(!isBuild)('direct', () => {
	test('Dummy.svelte?direct&svelte&type=style&sourcemap&lang.css', async () => {
		const response = await fetchFromPage(
			'src/Dummy.svelte?direct&svelte&type=style&sourcemap&lang.css',
			{
				headers: { Accept: 'text/css' }
			}
		);
		expect(response.ok).toBe(true);
		expect(response.headers.get('Content-Type')).toBe('text/css');
		const css = await response.text();
		expect(css).toMatchSnapshot();
	});
	test('Dummy.svelte?direct&svelte&type=script&sourcemap&lang.js', async () => {
		const response = await fetchFromPage(
			'src/Dummy.svelte?direct&svelte&type=script&sourcemap&lang.js',
			{
				headers: { Accept: 'application/javascript' }
			}
		);
		expect(response.ok).toBe(true);
		expect(response.headers.get('Content-Type')).toBe('application/javascript');
		const js = await response.text();
		expect(normalizeSnapshot(js)).toMatchSnapshot();
	});
});

describe.runIf(!isBuild)('ssrLoadModule', () => {
	let vite: ViteDevServer;
	let ssrLoadDummy;
	beforeAll(async () => {
		vite = await createServer({
			root: testDir + '/',
			appType: 'custom',
			server: { middlewareMode: true, hmr: false }
		});
		// needed to init plugins
		await vite.pluginContainer.buildStart({});
		ssrLoadDummy = async (query) =>
			vite
				.ssrLoadModule('./src/Dummy.svelte' + query, { fixStacktrace: true })
				.then((m) => m.default?.code ?? m.default);
		return async () => {
			await vite.close();
			vite = null;
			ssrLoadDummy = null;
		};
	});

	test('?raw', async () => {
		const result = await ssrLoadDummy('?raw');
		expect(result).toMatchSnapshot();
	});
	test('?raw&svelte&type=preprocessed', async () => {
		const result = await ssrLoadDummy('?raw&svelte&type=preprocessed');
		expect(result).toMatchSnapshot();
	});
	test('?raw&svelte&type=script', async () => {
		const result = await ssrLoadDummy('?raw&svelte&type=script');
		expect(normalizeSnapshot(result)).toMatchSnapshot();
	});
	test('?raw&svelte&type=script&compilerOptions={"customElement":true}', async () => {
		const result = await ssrLoadDummy(
			'?raw&svelte&type=script&compilerOptions={"customElement":true}'
		);
		expect(normalizeSnapshot(result)).toMatchSnapshot();
	});
	test('?raw&svelte&type=style', async () => {
		const result = await ssrLoadDummy('?raw&svelte&type=style');
		expect(result).toMatchSnapshot();
	});
});
