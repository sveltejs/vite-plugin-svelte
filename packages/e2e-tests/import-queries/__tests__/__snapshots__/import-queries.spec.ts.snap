// Vitest Snapshot v1

exports[`direct > Dummy.svelte?direct&svelte&type=script 1`] = `
"/* src/Dummy.svelte generated by Svelte v3.53.1 */
import {
	SvelteComponent as SvelteComponent$,
	append as append$,
	attr as attr$,
	detach as detach$,
	element as element$,
	init as init$,
	insert as insert$,
	listen as listen$,
	noop as noop$,
	safe_not_equal as safe_not_equal$,
	set_data as set_data$,
	text as text$
} from \\"svelte/internal\\";

function create_fragment(ctx) {
	let button$;
	let t0$;
	let t1$;
	let t2$;
	let mounted;
	let dispose;

	return {
		c() {
			button$ = element$(\\"button\\");
			t0$ = text$(/*name*/ ctx[0]);
			t1$ = text$(\\" clicks: \\");
			t2$ = text$(/*clicks*/ ctx[1]);
			attr$(button$, \\"class\\", \\"svelte-d8vj6a\\");
		},
		m(target, anchor) {
			insert$(target, button$, anchor);
			append$(button$, t0$);
			append$(button$, t1$);
			append$(button$, t2$);

			if (!mounted) {
				dispose = listen$(button$, \\"click\\", /*click_handler$*/ ctx[2]);
				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if (dirty & /*name*/ 1) set_data$(t0$, /*name*/ ctx[0]);
			if (dirty & /*clicks*/ 2) set_data$(t2$, /*clicks*/ ctx[1]);
		},
		i: noop$,
		o: noop$,
		d(detaching) {
			if (detaching) detach$(button$);
			mounted = false;
			dispose();
		}
	};
}

function instance$($$self, $$props, $$invalidate) {
	let { name } = $$props;
	let clicks = 0;

	const click_handler$ = () => {
		$$invalidate(1, clicks++, clicks);
	};

	$$self.$$set = $$props => {
		if ('name' in $$props) $$invalidate(0, name = $$props.name);
	};

	return [name, clicks, click_handler$];
}

class Dummy$ extends SvelteComponent$ {
	constructor(options) {
		super();
		init$(this, options, instance$, create_fragment, safe_not_equal$, { name: 0 });
	}
}

export default Dummy$;"
`;

exports[`direct > Dummy.svelte?direct&svelte&type=style&lang.css 1`] = `"button.svelte-d8vj6a{color:#000099}"`;

exports[`raw > Dummy.svelte?raw 1`] = `
"<script lang=\\"ts\\">
	export let name: string;
	let clicks = 0;
</script>

<button
	on:click={() => {
		clicks++;
	}}>{name} clicks: {clicks}</button
>

<style lang=\\"scss\\">
	$blue: blue;
	button {
		color: darken($blue, 20);
	}
</style>
"
`;

exports[`raw > Dummy.svelte?raw&svelte&type=preprocessed 1`] = `
"<script lang=\\"ts\\">export let name;
let clicks = 0;
</script>

<button
	on:click={() => {
		clicks++;
	}}>{name} clicks: {clicks}</button
>

<style lang=\\"scss\\">button {
  color: #000099;
}</style>
"
`;

exports[`raw > Dummy.svelte?raw&svelte&type=script 1`] = `
"/* src/Dummy.svelte generated by Svelte v3.53.1 */
import {
	SvelteComponent as SvelteComponent$,
	append as append$,
	attr as attr$,
	detach as detach$,
	element as element$,
	init as init$,
	insert as insert$,
	listen as listen$,
	noop as noop$,
	safe_not_equal as safe_not_equal$,
	set_data as set_data$,
	text as text$
} from \\"svelte/internal\\";

function create_fragment(ctx) {
	let button$;
	let t0$;
	let t1$;
	let t2$;
	let mounted;
	let dispose;

	return {
		c() {
			button$ = element$(\\"button\\");
			t0$ = text$(/*name*/ ctx[0]);
			t1$ = text$(\\" clicks: \\");
			t2$ = text$(/*clicks*/ ctx[1]);
			attr$(button$, \\"class\\", \\"svelte-d8vj6a\\");
		},
		m(target, anchor) {
			insert$(target, button$, anchor);
			append$(button$, t0$);
			append$(button$, t1$);
			append$(button$, t2$);

			if (!mounted) {
				dispose = listen$(button$, \\"click\\", /*click_handler$*/ ctx[2]);
				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if (dirty & /*name*/ 1) set_data$(t0$, /*name*/ ctx[0]);
			if (dirty & /*clicks*/ 2) set_data$(t2$, /*clicks*/ ctx[1]);
		},
		i: noop$,
		o: noop$,
		d(detaching) {
			if (detaching) detach$(button$);
			mounted = false;
			dispose();
		}
	};
}

function instance$($$self, $$props, $$invalidate) {
	let { name } = $$props;
	let clicks = 0;

	const click_handler$ = () => {
		$$invalidate(1, clicks++, clicks);
	};

	$$self.$$set = $$props => {
		if ('name' in $$props) $$invalidate(0, name = $$props.name);
	};

	return [name, clicks, click_handler$];
}

class Dummy$ extends SvelteComponent$ {
	constructor(options) {
		super();
		init$(this, options, instance$, create_fragment, safe_not_equal$, { name: 0 });
	}
}

export default Dummy$;"
`;

exports[`raw > Dummy.svelte?raw&svelte&type=script&compileOptions={"customElement":true} 1`] = `
"/* src/Dummy.svelte generated by Svelte v3.53.1 */
import {
	SvelteElement as SvelteElement$,
	append as append$,
	attribute_to_object as attribute_to_object$,
	detach as detach$,
	element as element$,
	flush as flush$,
	init as init$,
	insert as insert$,
	listen as listen$,
	noop as noop$,
	safe_not_equal as safe_not_equal$,
	set_data as set_data$,
	text as text$
} from \\"svelte/internal\\";

function create_fragment(ctx) {
	let button$;
	let t0$;
	let t1$;
	let t2$;
	let mounted;
	let dispose;

	return {
		c() {
			button$ = element$(\\"button\\");
			t0$ = text$(/*name*/ ctx[0]);
			t1$ = text$(\\" clicks: \\");
			t2$ = text$(/*clicks*/ ctx[1]);
			this.c = noop$;
		},
		m(target, anchor) {
			insert$(target, button$, anchor);
			append$(button$, t0$);
			append$(button$, t1$);
			append$(button$, t2$);

			if (!mounted) {
				dispose = listen$(button$, \\"click\\", /*click_handler$*/ ctx[2]);
				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if (dirty & /*name*/ 1) set_data$(t0$, /*name*/ ctx[0]);
			if (dirty & /*clicks*/ 2) set_data$(t2$, /*clicks*/ ctx[1]);
		},
		i: noop$,
		o: noop$,
		d(detaching) {
			if (detaching) detach$(button$);
			mounted = false;
			dispose();
		}
	};
}

function instance$($$self, $$props, $$invalidate) {
	let { name } = $$props;
	let clicks = 0;

	const click_handler$ = () => {
		$$invalidate(1, clicks++, clicks);
	};

	$$self.$$set = $$props => {
		if ('name' in $$props) $$invalidate(0, name = $$props.name);
	};

	return [name, clicks, click_handler$];
}

class Dummy$ extends SvelteElement$ {
	constructor(options) {
		super();
		this.shadowRoot.innerHTML = \`<style>button{color:#000099}</style>\`;

		init$(
			this,
			{
				target: this.shadowRoot,
				props: attribute_to_object$(this.attributes),
				customElement: true
			},
			instance$,
			create_fragment,
			safe_not_equal$,
			{ name: 0 },
			null
		);

		if (options) {
			if (options.target) {
				insert$(options.target, this, options.anchor);
			}

			if (options.props) {
				this.$set(options.props);
				flush$();
			}
		}
	}

	static get observedAttributes() {
		return [\\"name\\"];
	}

	get name() {
		return this.$$.ctx[0];
	}

	set name(name) {
		this.$$set({ name });
		flush$();
	}
}

export default Dummy$;"
`;

exports[`raw > Dummy.svelte?raw&svelte&type=style 1`] = `"button.svelte-d8vj6a{color:#000099}"`;

exports[`ssrLoadModule > ?raw 1`] = `
"<script lang=\\"ts\\">
	export let name: string;
	let clicks = 0;
</script>

<button
	on:click={() => {
		clicks++;
	}}>{name} clicks: {clicks}</button
>

<style lang=\\"scss\\">
	$blue: blue;
	button {
		color: darken($blue, 20);
	}
</style>
"
`;

exports[`ssrLoadModule > ?raw&svelte&type=preprocessed 1`] = `
"<script lang=\\"ts\\">export let name;
let clicks = 0;
</script>

<button
	on:click={() => {
		clicks++;
	}}>{name} clicks: {clicks}</button
>

<style lang=\\"scss\\">button {
  color: #000099;
}</style>
"
`;

exports[`ssrLoadModule > ?raw&svelte&type=script 1`] = `
"/* temp/serve/import-queries/src/Dummy.svelte generated by Svelte v3.53.1 */
import {
	SvelteComponent as SvelteComponent$,
	append as append$,
	attr as attr$,
	detach as detach$,
	element as element$,
	init as init$,
	insert as insert$,
	listen as listen$,
	noop as noop$,
	safe_not_equal as safe_not_equal$,
	set_data as set_data$,
	text as text$
} from \\"svelte/internal\\";

function create_fragment(ctx) {
	let button$;
	let t0$;
	let t1$;
	let t2$;
	let mounted;
	let dispose;

	return {
		c() {
			button$ = element$(\\"button\\");
			t0$ = text$(/*name*/ ctx[0]);
			t1$ = text$(\\" clicks: \\");
			t2$ = text$(/*clicks*/ ctx[1]);
			attr$(button$, \\"class\\", \\"svelte-d8vj6a\\");
		},
		m(target, anchor) {
			insert$(target, button$, anchor);
			append$(button$, t0$);
			append$(button$, t1$);
			append$(button$, t2$);

			if (!mounted) {
				dispose = listen$(button$, \\"click\\", /*click_handler$*/ ctx[2]);
				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if (dirty & /*name*/ 1) set_data$(t0$, /*name*/ ctx[0]);
			if (dirty & /*clicks*/ 2) set_data$(t2$, /*clicks*/ ctx[1]);
		},
		i: noop$,
		o: noop$,
		d(detaching) {
			if (detaching) detach$(button$);
			mounted = false;
			dispose();
		}
	};
}

function instance$($$self, $$props, $$invalidate) {
	let { name } = $$props;
	let clicks = 0;

	const click_handler$ = () => {
		$$invalidate(1, clicks++, clicks);
	};

	$$self.$$set = $$props => {
		if ('name' in $$props) $$invalidate(0, name = $$props.name);
	};

	return [name, clicks, click_handler$];
}

class Dummy$ extends SvelteComponent$ {
	constructor(options) {
		super();
		init$(this, options, instance$, create_fragment, safe_not_equal$, { name: 0 });
	}
}

export default Dummy$;"
`;

exports[`ssrLoadModule > ?raw&svelte&type=script&compileOptions={"customElement":true} 1`] = `
"/* temp/serve/import-queries/src/Dummy.svelte generated by Svelte v3.53.1 */
import {
	SvelteElement as SvelteElement$,
	append as append$,
	attribute_to_object as attribute_to_object$,
	detach as detach$,
	element as element$,
	flush as flush$,
	init as init$,
	insert as insert$,
	listen as listen$,
	noop as noop$,
	safe_not_equal as safe_not_equal$,
	set_data as set_data$,
	text as text$
} from \\"svelte/internal\\";

function create_fragment(ctx) {
	let button$;
	let t0$;
	let t1$;
	let t2$;
	let mounted;
	let dispose;

	return {
		c() {
			button$ = element$(\\"button\\");
			t0$ = text$(/*name*/ ctx[0]);
			t1$ = text$(\\" clicks: \\");
			t2$ = text$(/*clicks*/ ctx[1]);
			this.c = noop$;
		},
		m(target, anchor) {
			insert$(target, button$, anchor);
			append$(button$, t0$);
			append$(button$, t1$);
			append$(button$, t2$);

			if (!mounted) {
				dispose = listen$(button$, \\"click\\", /*click_handler$*/ ctx[2]);
				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if (dirty & /*name*/ 1) set_data$(t0$, /*name*/ ctx[0]);
			if (dirty & /*clicks*/ 2) set_data$(t2$, /*clicks*/ ctx[1]);
		},
		i: noop$,
		o: noop$,
		d(detaching) {
			if (detaching) detach$(button$);
			mounted = false;
			dispose();
		}
	};
}

function instance$($$self, $$props, $$invalidate) {
	let { name } = $$props;
	let clicks = 0;

	const click_handler$ = () => {
		$$invalidate(1, clicks++, clicks);
	};

	$$self.$$set = $$props => {
		if ('name' in $$props) $$invalidate(0, name = $$props.name);
	};

	return [name, clicks, click_handler$];
}

class Dummy$ extends SvelteElement$ {
	constructor(options) {
		super();
		this.shadowRoot.innerHTML = \`<style>button{color:#000099}</style>\`;

		init$(
			this,
			{
				target: this.shadowRoot,
				props: attribute_to_object$(this.attributes),
				customElement: true
			},
			instance$,
			create_fragment,
			safe_not_equal$,
			{ name: 0 },
			null
		);

		if (options) {
			if (options.target) {
				insert$(options.target, this, options.anchor);
			}

			if (options.props) {
				this.$set(options.props);
				flush$();
			}
		}
	}

	static get observedAttributes() {
		return [\\"name\\"];
	}

	get name() {
		return this.$$.ctx[0];
	}

	set name(name) {
		this.$$set({ name });
		flush$();
	}
}

export default Dummy$;"
`;

exports[`ssrLoadModule > ?raw&svelte&type=style 1`] = `"button.svelte-d8vj6a{color:#000099}"`;
